### PR TITLE
Updating error status on NIMcache object for reconcile failure

### PR DESCRIPTION
**Pull request by visheshtanksale**

### CODE REVIEW

**Review by slu2011 — APPROVED** [review]

**Comment by shivamerla** [comment]
@visheshtanksale as discussed, lets audit for spurious failures before we mark NIMService as Failed.

File: internal/controller/nimcache_controller.go [hunk 1]
```diff
@@ -815,6 +798,19 @@ func (r *NIMCacheReconciler) reconcileNIMCache(ctx context.Context, nimCache *ap
 		logger.Error(err, "reconciliation of caching job failed", "job", getJobName(nimCache))
 		return ctrl.Result{}, err
 	}
+
+	obj := &appsv1alpha1.NIMCache{}
```
Comment (shivamerla) [inline internal/controller/nimcache_controller.go:802]: Can we pull this into a separate function which does get and update of CR status.

File: internal/controller/nimcache_controller.go [hunk 2]
```diff
@@ -815,6 +798,19 @@ func (r *NIMCacheReconciler) reconcileNIMCache(ctx context.Context, nimCache *ap
...
+	errGet := r.Get(ctx, types.NamespacedName{Name: nimCache.Name, Namespace: nimCache.GetNamespace()}, obj)
+	if errGet != nil {
+		logger.Error(errGet, "error getting NIMCache", "name", nimCache.Name)
+		return ctrl.Result{}, errGet
+	}
+	obj.Status = nimCache.Status
```
[the remark this answers is not in the document]
Comment (shivamerla) [reply internal/controller/nimcache_controller.go:808]: nvm, i overlooked that we are copying the entire `status` object.

File: internal/controller/nimcache_controller.go [hunk 3]
```diff
@@ -153,11 +152,22 @@ func (r *NIMCacheReconciler) Reconcile(ctx context.Context, req ctrl.Request) (c
...
 	result, err := r.reconcileNIMCache(ctx, nimCache)
 	if err != nil {
+		obj := &appsv1alpha1.NIMCache{}
+		errGet := r.Get(ctx, types.NamespacedName{Name: nimCache.Name, Namespace: nimCache.GetNamespace()}, obj)
+		if errGet != nil {
+			logger.Error(errGet, "error getting NIMCache", "name", nimCache.Name)
+			return result, errGet
+		}
 		logger.Error(err, "error reconciling NIMCache", "name", nimCache.Name)
+		conditions.UpdateCondition(&obj.Status.Conditions, appsv1alpha1.NimCacheConditionPVCCreated, metav1.ConditionFalse, "Failed", err.Error())
```
[the remark this answers is not in the document]
Comment (visheshtanksale) [reply internal/controller/nimcache_controller.go]: Fixed this added a new condition.

[the remark this answers is not in the document]
Comment (shivamerla) [reply internal/controller/nimcache_controller.go]: looks good

File: internal/controller/nimcache_controller.go [hunk 4]
```diff
@@ -153,14 +152,19 @@ func (r *NIMCacheReconciler) Reconcile(ctx context.Context, req ctrl.Request) (c
 			return ctrl.Result{}, nil
 		}
 	}
-
 	// Handle nim-cache reconciliation
 	result, err := r.reconcileNIMCache(ctx, nimCache)
 	if err != nil {
 		logger.Error(err, "error reconciling NIMCache", "name", nimCache.Name)
+		conditions.UpdateCondition(&nimCache.Status.Conditions, appsv1alpha1.NimCacheConditionReconcileFailed, metav1.ConditionTrue, "Error", err.Error())
```
Comment (shivamerla) [inline internal/controller/nimcache_controller.go:159]:
```suggestion
		conditions.UpdateCondition(&nimCache.Status.Conditions, appsv1alpha1.NimCacheConditionReconcileFailed, metav1.ConditionTrue, "ReconcileFailed", err.Error())
```

File: internal/controller/nimcache_controller.go [hunk 5]
```diff
@@ -815,9 +792,33 @@ func (r *NIMCacheReconciler) reconcileNIMCache(ctx context.Context, nimCache *ap
 		logger.Error(err, "reconciliation of caching job failed", "job", getJobName(nimCache))
 		return ctrl.Result{}, err
 	}
+
+	conditions.IfPresentUpdateCondition(&nimCache.Status.Conditions, appsv1alpha1.NimCacheConditionReconcileFailed, metav1.ConditionTrue, "Succeed", "")
```
Comment (shivamerla) [inline internal/controller/nimcache_controller.go:796]:
```suggestion
	conditions.IfPresentUpdateCondition(&nimCache.Status.Conditions, appsv1alpha1.NimCacheConditionReconcileFailed, metav1.ConditionFalse, "ReconcileFailed", "")
```

Comment (visheshtanksale) [reply internal/controller/nimcache_controller.go]: Fixed

**Review by shivamerla — APPROVED** [review]
lgtm

**Pull request merged**